pkg/controller/cyclenoderequest/transitioner/healthchecks.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	return renderedEndpoint.String(), nil
+}
+
+// healthCheckPassed checks if the statusCode returned matches the set of valid status code for the health check
+// as well as if the body matches regex provided
+func healthCheckPassed(healthCheck v1.HealthCheck, statusCode uint, body []byte) (bool, string) {
+	// If there is not regex match string specified, don't check the body of the response
+	r := regexp.MustCompile(healthCheck.RegexMatch)
+	if healthCheck.RegexMatch != "" && !r.Match(body) {
+		return false, fmt.Sprintf("regex %s did not match body %b", healthCheck.RegexMatch, body)
+	}
+
+	for _, validStatusCode := range healthCheck.ValidStatusCodes {
+		if statusCode == validStatusCode {
+			return true, ""
+		}
+	}
+
+	return false, fmt.Sprintf("status code %d returned, did not match expected %v", statusCode, healthCheck.ValidStatusCodes)
+}
+
+// performHealthCheck makes the health check request to the endpoint specified, reads the body and returns
+// the status code/body to determinate weather it passed
+func (t *CycleNodeRequestTransitioner) performHealthCheck(endpoint string) (uint, []byte, error) {
+	resp, err := t.rm.HttpClient.Get(endpoint)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	defer resp.Body.Close()
+
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	return uint(resp.StatusCode), bytes, nil
+}
+
+// performInitialHealthChecks on the nodes selected to be terminated before cycling begin. If any health
+// check fails return an error to prevent cycling from starting
+func (t *CycleNodeRequestTransitioner) performInitialHealthChecks(kubeNodes []corev1.Node) error {
+	t.cycleNodeRequest.Status.HealthChecks = make(map[string]v1.HealthCheckStatus)
+
+	// Build a set of ready nodes from which to check below
+	var readyNodesSet = make(map[string]corev1.Node)
+
+	// Collect all nodes in the nodegroup and assign Skip=true
+	// Health checks will not be performed on them during cycling in the ScalingUp phase
+	for _, kubeNode := range kubeNodes {
+		// Do not use the provider assigned name of the node
+		// If the name of the node is the private dns address, there is a risk of new instances being
+		// assigned the same address as a previously terminated node in the initial set in the nodegroup.
+		// This would cause the new node health checks to be skipped
+		nodeName := fmt.Sprintf("%s/%s", kubeNode.Spec.ProviderID, kubeNode.Name)
+		readyNodesSet[nodeName] = kubeNode
+
+		// Set up the health check statuses for each node
+		// All nodes present in the nodegroup before cycling should be skipped after cycling has begun
+		t.cycleNodeRequest.Status.HealthChecks[nodeName] = v1.HealthCheckStatus{
+			Skip: true,
+		}
+	}
+
+	// Perform healthchecks on all nodes in the nodegroup before cycling begins
+	// If any healthchecks fail, cycling should not start
+	for _, node := range t.cycleNodeRequest.Status.NodesToTerminate {
+		nodeName := fmt.Sprintf("%s/%s", node.ProviderID, node.Name)
```

Comment:
this hash function is used multiple times across functions. Break out into its own function so the hash encode format is in one place